src/cmd/version.go

```diff
@@ -1,24 +1,104 @@
 package cmd
 
 import (
+	"encoding/json"
 	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"runtime"
 
 	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
 )
 
-var version = "development"
+type Build struct {
+	Version         string          `json:"version,omitempty"`
+	Commit          string          `json:"git,omitempty"`
+	GoInfo          GoInfo          `json:"go,omitempty"`
+	OpslevelVersion OpslevelVersion `json:"opslevel,omitempty"`
+}
+
+type OpslevelVersion struct {
+	Commit  string `json:"app_commit,omitempty"`
+	Version string `json:"app_version,omitempty"`
+}
+
+type GoInfo struct {
+	Version  string `json:"version,omitempty"`
+	Compiler string `json:"compiler,omitempty"`
+	OS       string `json:"os,omitempty"`
+	Arch     string `json:"arch,omitempty"`
+}
+
+var (
+	version = "development"
+	commit  = "none"
+	build   Build
+)
+
+func initBuild() {
+	build.Version = version
+	if len(commit) >= 12 {
+		build.Commit = commit[:12]
+	} else {
+		build.Commit = commit
+	}
+
+	build.GoInfo = getGoInfo()
+	build.OpslevelVersion = getOpslevelVersion()
+}
+
+func getGoInfo() GoInfo {
+	return GoInfo{
+		Version:  runtime.Version(),
+		Compiler: runtime.Compiler,
+		OS:       runtime.GOOS,
+		Arch:     runtime.GOARCH,
+	}
+}
+
+func getOpslevelVersion() OpslevelVersion {
+	// Need to update all of this when we switch over to resty client
+	apiUrl := viper.GetString("base-url") + "/api/ping"
+	response, err := http.Get(apiUrl)
+	if err != nil {
+		fmt.Print(err.Error())
+		os.Exit(1)
+	}
+
+	responseData, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var opslevelVersion OpslevelVersion
+	json.Unmarshal(responseData, &opslevelVersion)
+	if len(opslevelVersion.Commit) >= 12 {
+		opslevelVersion.Commit = opslevelVersion.Commit[:12]
+	}
+
+	return opslevelVersion
+}
 
 var versionCmd = &cobra.Command{
 	Use:   "version",
 	Short: "Print version information",
 	Long:  `Print version information`,
-	Run:   runVersion,
+	RunE:  runVersion,
```

Comment:
ouu fancy - does this just do cobra checkErr automatically?

Reply:
So this is what replicated had in their command and I updated it to that because Run doesn't accept a function that returns an error.

Reply:
Nice - we could reduce alot of code if we refactor for this - i'll cut a backlog ticket